File: clients/ruby/sym.rb
# sym — the SymBridge client for Ruby.
#
#     require_relative "sym"
#     mean = Sym.call("python", "statistics.mean", [[1, 2, 3]])
#     lst  = Sym.new_obj("java", "java.util.ArrayList")
#     lst.add("from ruby")
#     puts lst.size
#
# Ruby importing Python libs. Java objects living in Ruby variables.
# The hub is a Sym stdio host this client spawns and owns.
require "json"
require "open3"

module Sym
  @stdin = @stdout = @wait = nil
  @next_id = 0

  def self.root
    return ENV["SYM_HOME"] if ENV["SYM_HOME"] && File.directory?(File.join(ENV["SYM_HOME"], "bridge"))
    home = File.join(Dir.home, ".sym")
    return home if File.directory?(File.join(home, "bridge"))
    d = File.expand_path(File.dirname(__FILE__))
    5.times do
      return d if File.directory?(File.join(d, "bridge"))
      d = File.dirname(d)
    end
    core = `python3 -c "import sym_lang, os; print(os.path.join(os.path.dirname(sym_lang.__file__), 'core'))" 2>/dev/null`.strip
    return core if !core.empty? && File.directory?(File.join(core, "bridge"))
    raise "sym: no Sym found (pip install sym-lang, or set SYM_HOME)"
  end

  def self.ensure_host
    return if @stdin
    @stdin, @stdout, @wait = Open3.popen2("python3", File.join(root, "bridge", "stdio_host.py"))
    at_exit { close }
  end

  def self.request(msg)
    ensure_host
    @next_id += 1
    msg["id"] = @next_id
    @stdin.puts(JSON.generate(msg))
    @stdin.flush
    loop do
      line = @stdout.gets or raise "sym host died"
      resp = JSON.parse(line) rescue next
      next unless resp["id"] == @next_id
      raise "sym: #{resp["error"]}" unless resp["ok"]
      return wrap(resp["value"])
    end
  end

  def self.wrap(v)
    if v.is_a?(Hash) && v["__sym__"] == "handle"
      Handle.new(v)
    elsif v.is_a?(Array)
      v.map { |x| wrap(x) }
    else
      v
    end
  end

  def self.unwrap(args)
    args.map { |a| a.is_a?(Handle) ? a.tag : a }
  end

  def self.import(lang, target) = request({ "op" => "import", "lang" => lang, "target" => target })
  def self.call(lang, target, args = []) = request({ "op" => "call", "lang" => lang, "target" => target, "args" => unwrap(args) })
  def self.new_obj(lang, target, args = []) = request({ "op" => "new", "lang" => lang, "target" => target, "args" => unwrap(args) })
  def self.block(lang, code) = request({ "op" => "exec", "lang" => lang, "code" => code })
  def self.get(key) = request({ "op" => "get", "key" => key })
  def self.set(key, value) = request({ "op" => "set", "key" => key, "value" => value })
  def self.c_call(lib_fn, args, ret = "int", argtypes = []) = request({ "op" => "call", "lang" => "c", "target" => lib_fn, "args" => args, "ret" => ret, "argtypes" => argtypes })

  def self.close
    return unless @stdin
    begin; request({ "op" => "shutdown" }); rescue; end
    begin; @stdin.close; rescue; end
    @stdin = nil
  end

  class Handle
    attr_reader :tag
    def initialize(tag); @tag = tag; end
    def method_missing(name, *args)
      Sym.request({ "op" => "hcall", "lang" => @tag["runtime"],
                    "handle" => @tag["id"], "method" => name.to_s,
                    "args" => Sym.unwrap(args) })
    end
    def respond_to_missing?(*); true; end
    def [](key)
      Sym.request({ "op" => "index", "lang" => @tag["runtime"],
                    "handle" => @tag["id"], "args" => [key] })
    end
    def to_s; "#<Sym::Handle #{@tag["runtime"]}##{@tag["id"]} #{@tag["type"]}>"; end
  end
end

# ── selftest: this consumer against every provider ──────────
if __FILE__ == $0
  checks = {
    "java"   => -> { Sym.call("java", "java.lang.Math.pow", [2, 5]) == 32.0 },
    "js"     => -> { Sym.call("js", "Math.max", [3, 9, 2]) == 9 },
    "python" => -> { Sym.call("python", "math.sqrt", [81]) == 9.0 },
    "php"    => -> { Sym.call("php", "strtoupper", ["sym"]) == "SYM" },
    "ruby"   => -> { Sym.call("ruby", "Math.sqrt", [144]) == 12.0 },
    "r"      => -> { Sym.call("r", "mean", [[1, 2, 3, 4, 5]]) == 3 },
    "perl"   => -> { Sym.import("perl", "POSIX"); Sym.call("perl", "POSIX.floor", [3.7]) == 3 },
    "c"      => -> { (Sym.c_call("m.sqrt", [2.0], "double") - 1.41421).abs < 0.001 },
  }
  checks.each { |name, fn| puts "  #{fn.call ? "✅" : "❌"} ruby → #{name}" }
  lst = Sym.new_obj("java", "java.util.ArrayList")
  lst.add("from ruby")
  puts "  #{lst.size == 1 ? "✅" : "❌"} ruby → java live object"
  Sym.close
  puts "MATRIX_ROW_OK ruby"
end

File: clients/perl/sym.pl
#!/usr/bin/perl
# sym — the SymBridge client for Perl.
#     require "sym.pl";
#     print Sym::call("python", "math.sqrt", [81]);   # Perl importing Python
package Sym;
use strict; use warnings;
use JSON::PP;
use IPC::Open2;
use File::Basename qw(dirname);

my $json = JSON::PP->new->utf8->allow_nonref;
my ($OUT, $IN, $PID);
my $ID = 0;

sub root {
    return $ENV{SYM_HOME} if $ENV{SYM_HOME} && -d "$ENV{SYM_HOME}/bridge";
    my $home = "$ENV{HOME}/.sym";
    return $home if -d "$home/bridge";
    my $d = dirname(__FILE__);
    for (1..5) {
        return $d if -d "$d/bridge";
        $d = dirname($d);
    }
    my $core = `python3 -c "import sym_lang, os; print(os.path.join(os.path.dirname(sym_lang.__file__), 'core'))" 2>/dev/null`;
    chomp $core;
    return $core if $core && -d "$core/bridge";
    die "sym: no Sym found (pip install sym-lang, or set SYM_HOME)\n";
}

sub ensure {
    return if $PID;
    my $r = root();
    $PID = open2($IN, $OUT, "python3", "$r/bridge/stdio_host.py");
}

sub request {
    my ($msg) = @_;
    ensure();
    $msg->{id} = ++$ID;
    print $OUT $json->encode($msg), "\n";
    $OUT->flush;
    while (my $line = <$IN>) {
        my $resp = eval { $json->decode($line) } or next;
        next unless ($resp->{id} // -1) == $ID;
        die "sym: $resp->{error}\n" unless $resp->{ok};
        return wrap($resp->{value});
    }
    die "sym host died\n";
}

sub wrap {
    my ($v) = @_;
    if (ref($v) eq "HASH" && ($v->{"__sym__"} // "") eq "handle") {
        return Sym::Handle->new($v);
    }
    return [map { wrap($_) } @$v] if ref($v) eq "ARRAY";
    return $v;
}

sub unwrap { [map { ref($_) eq "Sym::Handle" ? $_->{tag} : $_ } @{$_[0] // []}] }

sub simport { request({ op => "import", lang => $_[0], target => $_[1] }) }
sub call    { request({ op => "call", lang => $_[0], target => $_[1], args => unwrap($_[2]) }) }
sub new_obj { request({ op => "new", lang => $_[0], target => $_[1], args => unwrap($_[2]) }) }
sub block   { request({ op => "exec", lang => $_[0], code => $_[1] }) }
sub c_call  { request({ op => "call", lang => "c", target => $_[0], args => $_[1], ret => $_[2] // "int", argtypes => $_[3] // [] }) }
sub close_  { eval { request({ op => "shutdown" }) }; }

package Sym::Handle;
our $AUTOLOAD;
sub new { my ($c, $tag) = @_; bless { tag => $tag }, $c }
sub AUTOLOAD {
    my $self = shift;
    (my $m = $AUTOLOAD) =~ s/.*:://;
    return if $m eq "DESTROY";
    Sym::request({ op => "hcall", lang => $self->{tag}{runtime},
                   handle => $self->{tag}{id}, method => $m,
                   args => Sym::unwrap(\@_) });
}

package main;
unless (caller) {
    my @checks = (
        ["java",   sub { Sym::call("java", "java.lang.Math.pow", [2, 5]) == 32 }],
        ["js",     sub { Sym::call("js", "Math.max", [3, 9, 2]) == 9 }],
        ["python", sub { Sym::call("python", "math.sqrt", [81]) == 9 }],
        ["php",    sub { Sym::call("php", "strtoupper", ["sym"]) eq "SYM" }],
        ["ruby",   sub { Sym::call("ruby", "Math.sqrt", [144]) == 12 }],
        ["r",      sub { Sym::call("r", "mean", [[1, 2, 3, 4, 5]]) == 3 }],
        ["perl",   sub { Sym::simport("perl", "POSIX"); Sym::call("perl", "POSIX.floor", [3.7]) == 3 }],
        ["c",      sub { abs(Sym::c_call("m.sqrt", [2.0], "double", ["double"]) - 1.41421) < 0.001 }],
    );
    for my $c (@checks) {
        print "  ", ($c->[1]->() ? "\x{2705}" : "\x{274c}"), " perl \x{2192} $c->[0]\n";
    }
    my $lst = Sym::new_obj("java", "java.util.ArrayList");
    $lst->add("from perl");
    print "  ", ($lst->size() == 1 ? "\x{2705}" : "\x{274c}"), " perl \x{2192} java live object\n";
    Sym::close_();
    print "MATRIX_ROW_OK perl\n";
}
1;
